[GUI][CGUIAction] Remove friend class, create proper interface

enen92 opened the pull request:
## Description
This PR provides a simple cleanup of the `CGUIAction` class by removing the friend class `CGUIControlFactory` - seems an anti-pattern to me since this class is actually just a container and according to the comment, the friend class was just added to avoid creating getters/setters (the fastest path). Also applies a few changes related to our codeguidelines.

## Motivation and context
I want to reuse CGUIAction in skin timers to solve a few inconsistencies of builtin execution via skintimers vs when the same action is registered in a given window. Adding yet another friend class to the container doesn't sound the right path, specially considering the type of mutations around CGUIAction and the fact all will run on the render thread anyway.

## How has this been tested?
Compile and runtime test, checking actions are correctly executed in windows.

## What is the effect on users?
None, if any it's my fault :)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

enen92 committed:
[GUI][CGUIAction] Remove friend class, create proper interface

enen92 commented:
Well I tried not to ask for your review this time @howie-f but I got no response from anyone :)

howie-f commented on xbmc/guilib/GUIAction.h:
```diff
@@ -44,17 +55,20 @@ class CGUIAction
+   * Configure CGUIAction to send threaded messages
+   */
+  void SetSendThreadMessages(bool sendThreadMessages);
+  /**
+   * Prune any action pairs stored in the CGUIAction
+   */
+  void Append(const CondActionPair& pair);
```
this doxy doesnt seem correct

enen92 replied:
nice find, c/p mistake :)

howie-f reviewed (approved):
L almost GTM

ksooo commented on xbmc/guilib/GUIAction.h:
```diff
@@ -21,21 +21,32 @@ class CGUIListItem; typedef std::shared_ptr<CGUIListItem> CGUIListItemPtr;
+  /**
+   * Pair containing the info condition (as string) and the action (as string)
+   */
+  struct CondActionPair
```
I would not name the type `Pair`. If we decide later to add attribute(s), we have to rename the type. Naming the type `Pair`somehow exposes an implementation detail.

ksooo commented on xbmc/guilib/GUIControlFactory.cpp:
```diff
@@ -492,20 +492,20 @@ bool CGUIControlFactory::GetAnimations(TiXmlNode *control, const CRect &rect, in
     if (pElement->FirstChild())
     {
-      CGUIAction::cond_action_pair pair;
+      CGUIAction::CondActionPair pair;
       pair.condition = XMLUtils::GetAttribute(pElement, "condition");
       pair.action = pElement->FirstChild()->Value();
```
Add a constructor to `CGUIAction::ConditionPair`, taking the condition + action ?

ksooo commented on xbmc/guilib/GUIControlFactory.cpp:
```diff
@@ -922,7 +922,8 @@ CGUIControl* CGUIControlFactory::Create(int parentID, const CRect &rect, TiXmlEl
   GetActions(pControlNode, "onfocus", focusActions);
   GetActions(pControlNode, "onunfocus", unfocusActions);
-  focusActions.m_sendThreadMessages = unfocusActions.m_sendThreadMessages = true;
+  focusActions.SetSendThreadMessages(true);
+  unfocusActions.SetSendThreadMessages(true);
```
As far as I can see this function will never be called with `false`. So, why not remove the bool parameter and define the semantics of the function to set the "SetSendThreadMessages" property to true.

enen92 commented:
Thanks for the comments, I think I addressed them all. Took the constructor opportunity to also to implement getters and setters for the executable action

ksooo commented on xbmc/guilib/GUIAction.h:
```diff
@@ -16,26 +16,74 @@ class CGUIControl;
+  /**
+   * Structure containing the the executable action definition
+   */
+  struct ExecutableAction
```
Should be a `class` - you fully encapsulated members, which is good.

ksooo replied:
What? Not really, see https://stackoverflow.com/questions/92859/what-are-the-differences-between-struct-and-class-in-c

ksooo replied:
What? Not really, see https://stackoverflow.com/questions/92859/what-are-the-differences-between-struct-and-class-in-c

howie-f commented on xbmc/guilib/GUIAction.cpp:
```diff
@@ -72,23 +98,43 @@ void CGUIAction::SetNavigation(int id)
     }
   }
   m_actions.emplace_back();
-  m_actions.back().action = std::move(strId);
+  m_actions.back().SetAction(std::move(strId));
```
just do `m_actions.emplace_back(strId, std::string());` then we don't need the default ctor

ksooo replied:
Imo, this std::move buzz is nonsense for the use cases we use it. Are those function really performance/ressource saving critical? Just as a side note.

ksooo commented on xbmc/guilib/GUIAction.h:
```diff
@@ -16,26 +16,74 @@ class CGUIControl;
+    /**
+    * Executable action default constructor
+    */
+    ExecutableAction() = default;
```
Why? Does the instance not work in case somebody prefers to use default ctor and dedicated setters? `delete`should only be used if the instance does not work when constructed this way.

ksooo replied:
> although struct and class are equal

What? Not really, see https://stackoverflow.com/questions/92859/what-are-the-differences-between-struct-and-class-in-c

ksooo replied:
>  although struct and class are equal

What? Not really, see https://stackoverflow.com/questions/92859/what-are-the-differences-between-struct-and-class-in-c

ksooo commented on xbmc/guilib/GUIAction.cpp:
```diff
@@ -55,10 +81,10 @@ int CGUIAction::GetNavigation() const
-      if (i.condition.empty() || infoMgr.EvaluateBool(i.condition, INFO::DEFAULT_CONTEXT))
-        return atoi(i.action.c_str());
+      if (!i.HasCondition() || infoMgr.EvaluateBool(i.GetCondition(), INFO::DEFAULT_CONTEXT))
+        return atoi(i.GetCondition().c_str());
```
while we are at it, `std::atoi`, for clarity

ksooo commented on xbmc/guilib/GUIAction.cpp:
```diff
@@ -72,23 +98,43 @@ void CGUIAction::SetNavigation(int id)
+
+void CGUIAction::Append(const ExecutableAction& pair)
+{
+  m_actions.emplace_back(pair);
```
+1 for changing pair to executableAction

ksooo replied:
`std::move` nonsense. ;-) And makes the code less readable.

ksooo reviewed (approved):
Looks good to me now.

enen92 commented:
Unfortunately there is something broken in this PR regarding navigation

ksooo commented on xbmc/guilib/GUIControlFactory.cpp:
```diff
@@ -490,22 +490,23 @@ bool CGUIControlFactory::GetAnimations(TiXmlNode *control, const CRect &rect, in
-      pair.condition = XMLUtils::GetAttribute(pElement, "condition");
-      pair.action = pElement->FirstChild()->Value();
-      action.m_actions.push_back(pair);
+      CGUIAction::CExecutableAction action = CGUIAction::CExecutableAction(
+          XMLUtils::GetAttribute(pElement, "condition"), pElement->FirstChild()->Value());
```
Shorter: `const CGUIAction::CExecutableAction action{XMLUtils::GetAttribute(pElement, "condition"), pElement->FirstChild()->Value()};`

enen92 replied:
This is just a style (apart from the missing const specifier) thing right? 
I don't see where the provided alternative is any different from what I'm already doing (hence I don't understand the `Prefer ctor with params over default ctor, followed by assigment` unless you're just providing an overall comment)

ksooo replied:
what you wrote means:

1) default construct `CGUIAction::CExecutableAction`
2) construct a CGUIAction::CExecutableAction with the given params
3) call (the compiler-generated) CGUIAction::CExecutableActionoperator= to copy the data from the second instance to the first

Yes, modern compilers can optimize this, but I tend never to rely on such magic.

What I propose menas:

1) construct a CGUIAction::CExecutableAction with the given params

done.

howie-f replied:
_confused_ 
i was of the strong opinion what enen92 had there is copy initialization (neither default ctor nor operator= involved, because it is not an assignment but an init) and your proposal is direct-list-initialization that falls back to the appropiate ctor (since we have no ctor `CGUIAction::CExecutableAction(std::initializer_list<>)`)...

ksooo replied:
> what enen92 had there is copy initialization

 my point is to use direct initialization, it's cheaper (and the code is very clear about what happens).

ksooo reviewed (approved):
Nitpicking.

enen92 commented on xbmc/guilib/GUIAction.cpp:
```diff
@@ -55,10 +81,10 @@ int CGUIAction::GetNavigation() const
-      if (i.condition.empty() || infoMgr.EvaluateBool(i.condition, INFO::DEFAULT_CONTEXT))
-        return atoi(i.action.c_str());
+      if (!i.HasCondition() || infoMgr.EvaluateBool(i.GetCondition(), INFO::DEFAULT_CONTEXT))
+        return std::atoi(i.GetCondition().c_str());
```
that's it! And I looked over this far too many times :)
working great now

enen92 commented:
Removed the default constructor on the last iteration, as pointed by @howie-f it doesn't make sense to have an action without an action. Also changed atoi to stoi to avoid the c_str.
Runtime tested.
